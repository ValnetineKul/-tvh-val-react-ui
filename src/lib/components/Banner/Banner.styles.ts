import { makeStyles } from '../../../themes/core';

const useStyles = makeStyles<
  void,
  'directionVertical' | 'directionHorizontal' | 'banner' | 'actionButton' | 'icon' | 'messageBoxHorizontal'
>({ name: 'Banner' })((theme, _, classes) => ({
  root: {
    [`&.${classes.directionHorizontal} .${classes.messageBoxHorizontal}`]: {
      display: 'flex',
      flex: 'auto',
      justifyContent: 'space-between',
    },
  },
  directionVertical: {},
  directionHorizontal: {},
  icon: {
    padding: theme.layoutSpacing['spacing/000'],
    opacity: 1,
    marginRight: theme.layoutSpacing['spacing/300'],
  },
  banner: {
    padding: `${theme.layoutSpacing['spacing/300']} ${theme.layoutSpacing['spacing/400']} !important`,
    borderRadius: 'inherit',
  },
  bannerMultiline: {
    [theme.breakpoints.up('xs')]: {
      alignItems: 'flex-start',
    },
  },
  message: {
    paddingBottom: theme.layoutSpacing['spacing/000'],
    display: 'flex',
    flex: 1,
  },
  messageBox: {
    [`& .${classes.actionButton}`]: {
      marginTop: theme.layoutSpacing['spacing/300'],
    },
  },
  messageBoxHorizontal: {},
  action: {
    gap: theme.layoutSpacing['spacing/300'],
  },
  actionButton: {},
  typography: {
    alignSelf: 'center',
    wordBreak: 'break-word',
    marginBottom: theme.layoutSpacing['spacing/200'],
  },
  fullWidth: {
    width: '100%',
  },
  surface: {
    borderRadius: theme.borderRadius['borderRadius/round/100'],
  },
}));

export default useStyles;
