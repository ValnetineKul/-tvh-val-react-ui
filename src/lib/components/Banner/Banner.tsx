import type { FC } from 'react';
import React, { useState, useEffect, useRef } from 'react';
import { capitalize } from '@mui/material/utils';
import Typography from '../Typography';
import type { BANNER_STATUSES } from './constants';
import useStyles from './Banner.styles';
import type { DataAttributes, ObjectValuesUnion } from '../../types/common';
import Alert from '../Alert';

type Status = ObjectValuesUnion<typeof BANNER_STATUSES>;

export interface BannerProps {
  status: Status;
  closable?: boolean;
  onClose?: () => void;
  className?: string;
  containerClassName?: string;
  message: string | React.ReactNode;
  button?: React.ReactElement;
  direction?: 'horizontal' | 'vertical';
  closeButtonProps?: DataAttributes;
}

const Banner: FC<BannerProps> = ({
  status,
  closable = false,
  onClose,
  className,
  containerClassName,
  message,
  button,
  direction = 'horizontal',
  closeButtonProps,
}) => {
  const { classes, cx } = useStyles();
  const [isMultiline, setIsMultiline] = useState(false);
  const messageRef = useRef<HTMLDivElement>(null);
  const directionClassName = (classes as Record<string, string>)[`direction${capitalize(direction)}`];

  useEffect(() => {
    const isBeingMultiline = () => {
      const isMulti =
        direction === 'horizontal' && messageRef && messageRef.current && messageRef.current.clientHeight > 24;
      setIsMultiline(!!isMulti);
    };

    isBeingMultiline();

    window.addEventListener('resize', isBeingMultiline);
    return () => {
      window.removeEventListener('resize', isBeingMultiline);
    };
  }, [message, direction]);

  const renderChildren = () => {
    if (typeof message === 'string') {
      return (
        <Typography className={classes.typography} component="div" ref={messageRef}>
          {message}
        </Typography>
      );
    }

    return <div ref={messageRef}>{message}</div>;
  };

  const renderTertiaryAction = () => {
    if (button) {
      return React.cloneElement(button, {
        size: 'sm',
        variant: 'tertiary',
        surfaceClassName: classes.actionButton,
      });
    }
    return null;
  };

  const renderAction = () => {
    return direction === 'horizontal' ? renderTertiaryAction() : undefined;
  };

  const renderMessage = () => {
    return (
      <div className={cx(classes.messageBox, classes.messageBoxHorizontal)}>
        {renderChildren()}
        {direction === 'vertical' && renderTertiaryAction()}
      </div>
    );
  };

  return (
    <Alert
      status={status}
      action={renderAction()}
      containerClassName={cx(classes.root, directionClassName, containerClassName)}
      className={cx(classes.banner, className, { [classes.bannerMultiline]: isMultiline })}
      classesOverrides={{ message: classes.message, action: classes.action }}
      surfaceClassName={classes.surface}
      message={renderMessage()}
      onClose={onClose}
      closable={closable}
      closeButtonProps={closeButtonProps}
    />
  );
};
export default Banner;
