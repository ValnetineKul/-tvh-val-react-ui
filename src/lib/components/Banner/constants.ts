export const BANNER_STATUSES = {
  SUCCESS: 'success',
  ERROR: 'error',
  WARNING: 'warning',
  INFO: 'info',
} as const;
