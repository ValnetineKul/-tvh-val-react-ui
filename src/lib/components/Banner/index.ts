import Banner from './Banner';

export { BANNER_STATUSES } from './constants';

export default Banner;
export type { BannerProps } from './Banner';
