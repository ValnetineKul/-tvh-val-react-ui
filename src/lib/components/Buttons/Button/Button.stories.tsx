import React from 'react';
import { createMeta, createTemplate } from '../../../story-utils';
import { buttonBaseStoryArgs } from '../../ButtonBase';
import { AngleLeft, AngleRight } from '../../Icon/icons/functional';
import Icon from '../../Icon';
import Button from './Button';

const Template = createTemplate(Button);

const Primitive = Template.bind({});

Primitive.args = {
  label: 'Label',
  variant: 'primary',
  type: 'button',
  size: 'md',
  fullWidth: false,
  disabled: false,
  isLoading: false,
  buttonProps: { 'data-testid': 'test' },
};

export const None = Template.bind({});
None.args = {
  ...Primitive.args,
};

None.storyName = 'none';

export const StartIcon = Template.bind({});
StartIcon.args = {
  ...Primitive.args,
  startIcon: <Icon icon={AngleLeft} />,
};

StartIcon.storyName = 'startIcon';

export const EndIcon = Template.bind({});
EndIcon.args = {
  ...Primitive.args,
  endIcon: <Icon icon={AngleRight} />,
};

EndIcon.storyName = 'endIcon';

export default createMeta({
  component: Button,
  title: 'Components/Buttons/Button',
  argTypes: {
    onClick: {
      action: { argTypesRegex: '^on.*' },
      control: { disable: true },
      table: {
        type: {
          summary: '(event: any) => void',
        },
      },
    },
    startIcon: {
      control: {
        disable: true,
      },
      table: {
        type: { summary: 'Icon with FontAwesome definition' },
      },
    },
    endIcon: {
      control: {
        disable: true,
      },
      table: {
        type: { summary: 'Icon FontAwesome definition' },
      },
    },
    ...buttonBaseStoryArgs,
  },
});
