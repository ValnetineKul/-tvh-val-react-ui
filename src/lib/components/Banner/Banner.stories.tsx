import { action } from '@storybook/addon-actions';
import React, { useState } from 'react';
import { createMeta, createTemplate } from '../../story-utils';
import Button from '../Buttons/Button';
import Collapse from '../Collapse';
import Typography from '../Typography';
import Banner from './Banner';

type Props = React.ComponentProps<typeof Banner>;
const Template = createTemplate(Banner);

export const Primitive = Template.bind({});
Primitive.args = {
  status: 'success',
  message: 'Banner message',
  button: <Button label="Label" onClick={action('onButtonClick')} variant="tertiary" />,
  closable: true,
};

Primitive.storyName = 'Banner';

const ExpandableTemplate = (args: Props) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <Banner
      {...args}
      message={
        <>
          <Typography>This is an infomessage.</Typography>
          <Collapse open={isOpen}>
            <Typography>
              Duis mollis, est non commodo luctus, nisi erat porttitor ligula, eget lacinia odio sem nec elit. Cras
              justo odio, dapibus ac facilisis in, egestas eget quam.
            </Typography>
          </Collapse>
          <Button
            label={isOpen ? 'Show less' : 'Show more'}
            variant="link"
            onClick={() => {
              setIsOpen(!isOpen);
            }}
          />
        </>
      }
    />
  );
};

export const Expandable = ExpandableTemplate.bind({});
Expandable.args = {
  status: 'info',
  button: <Button label="Label" onClick={action('onButtonClick')} variant="tertiary" />,
  closable: true,
};

Expandable.argTypes = {
  message: {
    control: { disable: true },
  },
};

Expandable.storyName = 'ExpandableBanner';

export default createMeta({
  component: Banner,
  title: 'Components/Banners/Banner',
  argTypes: {
    status: {
      options: ['success', 'error', 'warning', 'info'],
      control: {
        type: 'select',
      },
    },
    button: {
      control: { disable: true },
      table: {
        type: { summary: 'Tertiary Button' },
      },
    },
  },
  parameters: {
    docs: {
      transformSource: (src: string) => src.replace(/\[object Object\]/g, 'Typography'),
    },
  },
});
