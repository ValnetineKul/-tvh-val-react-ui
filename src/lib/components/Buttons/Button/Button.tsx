import { default as MuiButton } from '@mui/material/Button';
import type { ReactElement, MouseEventHandler, KeyboardEvent } from 'react';
import React, { forwardRef } from 'react';
import { capitalize } from '@mui/material/utils';
import type { ButtonBaseProps } from '../../ButtonBase';
import Surface from '../../Surfaces/Surface';
import type { Surface as SurfaceColor } from '../../../../themes/core';
import { useSurface } from '../../../../themes/core';
import SpinnerWithBackdrop from '../../ProgressIndicators/Spinner/SpinnerWithBackdrop';
import useStyles from './Button.styles';
import useGetButtonProps from './helpers/useGetButtonProps';
import type { DataAttributes } from '../../../types/common';

export interface ButtonCommonProps extends ButtonBaseProps {
  variant?: 'primary' | 'secondary' | 'tertiary' | 'link';
  type?: 'button' | 'submit';
  linkType?: 'default' | 'primary';
  label: string | ReactElement;
  size?: 'sm' | 'md' | 'lg';
  disabled?: boolean;
  isLoading?: boolean;
  fullWidth?: boolean;
  className?: string;
  surfaceClassName?: string;
  containerClassName?: string;
  onClick?: MouseEventHandler<HTMLButtonElement>;
  onKeyPress?: (event: KeyboardEvent<HTMLButtonElement>) => void;
  buttonProps?: DataAttributes;
}

interface NoIcon {
  startIcon?: never;
  endIcon?: never;
}
interface StartIcon {
  startIcon: ReactElement;
  endIcon?: never;
}
interface EndIcon {
  startIcon?: never;
  endIcon: ReactElement;
}

export type ButtonIconProps = NoIcon | StartIcon | EndIcon;

export type ButtonProps = ButtonCommonProps & ButtonIconProps;

const Button = forwardRef<HTMLButtonElement | HTMLAnchorElement, ButtonProps>(
  (
    {
      variant = 'primary',
      type = 'button',
      label,
      onClick,
      size = 'md',
      fullWidth = false,
      isLoading = false,
      className,
      surfaceClassName,
      containerClassName,
      ...props
    },
    ref
  ) => {
    const { startIcon, endIcon, linkType, buttonProps, ...restProps } = props;

    const { isLink, shouldShowLoader, buttonBaseProps } = useGetButtonProps(isLoading, restProps, variant);
    const shouldShowFullWidth = fullWidth && variant !== 'link';

    const { color: onSurface } = useSurface();

    let variantColor: SurfaceColor;
    switch (variant) {
      case 'primary':
        variantColor = 'Primary';
        break;
      case 'secondary':
        variantColor = 'Secondary';
        break;
      case 'tertiary':
        variantColor = onSurface;
        break;
      case 'link':
        variantColor = onSurface;
        break;
      default:
        variantColor = 'Primary';
    }

    const { classes, cx } = useStyles({ variantColor });
    const sizeClassName = variant !== 'link' && (classes as Record<string, string>)[`size${capitalize(size)}`];

    return (
      <Surface
        color={variantColor}
        className={cx(classes.surface, { [classes.fullWidth]: shouldShowFullWidth }, surfaceClassName)}
      >
        <SpinnerWithBackdrop
          className={cx(
            { [classes.tertiaryButtonBackdrop]: variant === 'tertiary', [classes.fullWidth]: shouldShowFullWidth },
            containerClassName
          )}
          isLoading={shouldShowLoader}
          isBackdropOpaque
        >
          <MuiButton
            type={isLink ? undefined : type}
            className={cx(
              classes.root,
              sizeClassName,
              {
                [classes.button]: variant !== 'link',
                [classes.tertiaryButton]: variant === 'tertiary',
                [classes.variantLink]: variant === 'link',
                [classes.linkTypeDefault]: variant === 'link' && linkType !== 'primary',
                [classes.linkTypePrimary]: variant === 'link' && linkType === 'primary',
              },
              className
            )}
            classes={{
              startIcon: classes.startIcon,
              endIcon: classes.endIcon,
            }}
            startIcon={startIcon}
            endIcon={endIcon}
            fullWidth={fullWidth}
            onClick={onClick}
            disableElevation
            ref={ref}
            {...restProps}
            {...buttonBaseProps}
            {...buttonProps}
          >
            {label}
          </MuiButton>
        </SpinnerWithBackdrop>
      </Surface>
    );
  }
);
Button.displayName = 'Button';

export default Button;
