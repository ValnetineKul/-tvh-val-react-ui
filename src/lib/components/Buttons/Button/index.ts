import Button from './Button';
import type { ButtonProps, ButtonCommonProps, ButtonIconProps } from './Button';

export default Button;
export type { ButtonProps, ButtonCommonProps, ButtonIconProps };
