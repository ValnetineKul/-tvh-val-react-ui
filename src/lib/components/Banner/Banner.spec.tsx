import React from 'react';
import { render } from '../../test-utils';
import Button from '../Buttons/Button/Button';
import Banner from './Banner';

describe('Banner', () => {
  it('Should render direction: horizontal correctly', () => {
    const { container } = render(<Banner status="info" message="Banner message" />);
    expect(container.firstChild).toMatchSnapshot('horizontal');
  });

  it('Should render direction: vertical correctly', () => {
    const { container } = render(<Banner status="info" direction="vertical" message="Banner message" />);
    expect(container.firstChild).toMatchSnapshot('vertical');
  });

  it('Should render custom component correctly', () => {
    const { container } = render(<Banner status="info" message={<div>Message</div>} />);
    expect(container.firstChild).toMatchSnapshot('customComponent');
  });

  it('Should have action button with function call', () => {
    const actionMock = jest.fn();
    const { getAllByRole } = render(
      <Banner
        status="info"
        closable
        button={<Button label="btn label" variant="primary" onClick={actionMock} />}
        message="Banner message"
      />
    );

    const actionButton = getAllByRole('button', { name: 'btn label' });
    expect(actionButton[0]).toBeInTheDocument();

    actionButton[0].click();
    expect(actionMock).toHaveBeenCalledTimes(1);
  });

  it('Should be hidden after hitting close button', async () => {
    jest.useFakeTimers();

    const { getByRole, container } = render(<Banner status="info" closable message="Banner message" />);

    const closeButton = getByRole('button', { name: 'close' });
    expect(closeButton).toBeInTheDocument();

    closeButton.click();
    jest.advanceTimersByTime(500);
    expect(container.firstChild).toHaveClass('MuiCollapse-hidden');
  });
});
