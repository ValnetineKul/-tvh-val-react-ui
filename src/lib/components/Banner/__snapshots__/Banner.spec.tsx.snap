// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner Should render custom component correctly: customComponent 1`] = `
<div
  class="MuiCollapse-root MuiCollapse-vertical tss-5-Banner-directionHorizontal-ref tss-tfyh55-Banner-root-Banner-directionHorizontal MuiCollapse-entered mui-pwcg7p-MuiCollapse-root"
  style="min-height: 0px;"
>
  <div
    class="MuiCollapse-wrapper MuiCollapse-vertical mui-smkl36-MuiCollapse-wrapper"
  >
    <div
      class="MuiCollapse-wrapperInner MuiCollapse-vertical mui-9l5vo-MuiCollapse-wrapperInner"
    >
      <div
        class="tss-xtcpa2-Surface-root-Banner-surface"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardInfo MuiAlert-standard tss-1t9irdj-Alert-root-Banner-banner mui-ig7n8t-MuiPaper-root-MuiAlert-root"
          role="alert"
        >
          <div
            class="tss-13p8oay-Alert-icon tss-10-Alert-icon-ref MuiAlert-icon mui-1ytlwq5-MuiAlert-icon"
          >
            <div
              class="tss-w1sjks-Icon-root-Icon-sizeMd"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
                color="inherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                infoCircle.svg
              </svg>
            </div>
          </div>
          <div
            class="tss-cdmgpl-Alert-message-Banner-message MuiAlert-message mui-acap47-MuiAlert-message"
          >
            <div
              class="tss-5-Banner-messageBoxHorizontal-ref tss-1512wcc-Banner-messageBox-Banner-messageBoxHorizontal"
            >
              <div>
                <div>
                  Message
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Banner Should render direction: horizontal correctly: horizontal 1`] = `
<div
  class="MuiCollapse-root MuiCollapse-vertical tss-5-Banner-directionHorizontal-ref tss-tfyh55-Banner-root-Banner-directionHorizontal MuiCollapse-entered mui-pwcg7p-MuiCollapse-root"
  style="min-height: 0px;"
>
  <div
    class="MuiCollapse-wrapper MuiCollapse-vertical mui-smkl36-MuiCollapse-wrapper"
  >
    <div
      class="MuiCollapse-wrapperInner MuiCollapse-vertical mui-9l5vo-MuiCollapse-wrapperInner"
    >
      <div
        class="tss-xtcpa2-Surface-root-Banner-surface"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardInfo MuiAlert-standard tss-1t9irdj-Alert-root-Banner-banner mui-ig7n8t-MuiPaper-root-MuiAlert-root"
          role="alert"
        >
          <div
            class="tss-13p8oay-Alert-icon tss-10-Alert-icon-ref MuiAlert-icon mui-1ytlwq5-MuiAlert-icon"
          >
            <div
              class="tss-w1sjks-Icon-root-Icon-sizeMd"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
                color="inherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                infoCircle.svg
              </svg>
            </div>
          </div>
          <div
            class="tss-cdmgpl-Alert-message-Banner-message MuiAlert-message mui-acap47-MuiAlert-message"
          >
            <div
              class="tss-5-Banner-messageBoxHorizontal-ref tss-1512wcc-Banner-messageBox-Banner-messageBoxHorizontal"
            >
              <div
                class="tss-4-Typography-weightRegular-ref tss-avim8z-Typography-root-Typography-weightRegular-Banner-typography-Typography-body500"
              >
                Banner message
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Banner Should render direction: vertical correctly: vertical 1`] = `
<div
  class="MuiCollapse-root MuiCollapse-vertical tss-1pivonf-Banner-root-Banner-directionVertical MuiCollapse-entered mui-pwcg7p-MuiCollapse-root"
  style="min-height: 0px;"
>
  <div
    class="MuiCollapse-wrapper MuiCollapse-vertical mui-smkl36-MuiCollapse-wrapper"
  >
    <div
      class="MuiCollapse-wrapperInner MuiCollapse-vertical mui-9l5vo-MuiCollapse-wrapperInner"
    >
      <div
        class="tss-xtcpa2-Surface-root-Banner-surface"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardInfo MuiAlert-standard tss-1t9irdj-Alert-root-Banner-banner mui-ig7n8t-MuiPaper-root-MuiAlert-root"
          role="alert"
        >
          <div
            class="tss-13p8oay-Alert-icon tss-10-Alert-icon-ref MuiAlert-icon mui-1ytlwq5-MuiAlert-icon"
          >
            <div
              class="tss-w1sjks-Icon-root-Icon-sizeMd"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
                color="inherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                infoCircle.svg
              </svg>
            </div>
          </div>
          <div
            class="tss-cdmgpl-Alert-message-Banner-message MuiAlert-message mui-acap47-MuiAlert-message"
          >
            <div
              class="tss-5-Banner-messageBoxHorizontal-ref tss-1512wcc-Banner-messageBox-Banner-messageBoxHorizontal"
            >
              <div
                class="tss-4-Typography-weightRegular-ref tss-avim8z-Typography-root-Typography-weightRegular-Banner-typography-Typography-body500"
              >
                Banner message
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
